### PR TITLE
[GC] Refactor the GC functions to clearly distinguish mark and sweep phases

### PR DESCRIPTION
**Pull request by agarwal-navin**
## Description
This change refactors the code around the main garbage collection run by separating it into 3 steps:
1. Pre GC step - This includes initializing from base state (if not already done) and notifying container runtime that GC is starting.
2. GC step - This includes 2 main parts - running mark phase and running sweep phase.
3. Post GC step - This includes clearing state and logging any events.

The goal is to clearly indicate what logic runs in which step and also consolidate the logic of mark and sweep phases.

## Reviewer Guidance
This is a refactor. There is no change in functionality.

### CODE REVIEW

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 1]
```diff
@@ -482,53 +490,172 @@ export class GarbageCollector implements IGarbageCollector {
...
+
+	/**
+	 * Runs any steps after GC completes. It logs unreferenced events, increments counters and does other cleanup.
+	 */
+	private async runPostGCSteps(logger: ITelemetryLogger) {
+		// Log pending unreferenced events such as a node being used after inactive. This is done after GC runs and
+		// updates its state so that we don't send false positives based on intermediate state. For example, we may get
+		// reference to an unreferenced node from another unreferenced node which means the node wasn't revived.
+		await this.telemetryTracker.logPendingEvents(logger);
+		this.newReferencesSinceLastRun.clear();
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:566]: This moved from middle of Mark step to the end. Glancing at the code it seems fine, just curious if it was necessary or just seemed cleaner.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 2]
```diff
@@ -537,30 +664,145 @@ export class GarbageCollector implements IGarbageCollector {
...
+		 * Sweep mode - Sweep ready modes are deleted.
+		 *
+		 * These modes serve as staging for applications that want to enable sweep by providing an incremental
+		 * way to test and validate sweep works as expected.
+		 */
+		if (this.configs.testMode) {
+			// If we are running in GC test mode, unreferenced nodes (gcResult.deletedNodeIds) are deleted.
 			this.runtime.updateUnusedRoutes(gcResult.deletedNodeIds);
-		} else if (this.configs.tombstoneMode) {
+			return gcData;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:679]: Do we not need to update gcData here?

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:566]: It wasn't necessary. It's just cleaner. It needs to reset after GC completes.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:679]: Not with how test mode currently works. I am assuming you mean updating gcData similar to how it happens with actual sweep.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 3]
```diff
@@ -537,30 +664,145 @@ export class GarbageCollector implements IGarbageCollector {
...
+				// Stop tracking so as to clear out any running timers.
+				nodeStateTracker.stopTracking();
+				// Delete the node as we don't need to track it any more.
+				this.unreferencedNodesState.delete(nodeId);
+			}
+			// TODO: GC:Validation - assert that the deleted node is not a duplicate
+			this.deletedNodes.add(nodeId);
+		}
+
+		// 3. Update and return the passed in gcData by removing deleted nodes from it.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:713]: I'd rephrase this, you're not modifying the passed in gcData.  Also made me consider a slight refactor which I'm trying locally and will run by you if I like it

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 4]
```diff
@@ -482,53 +490,172 @@ export class GarbageCollector implements IGarbageCollector {
...
 			},
 			{ end: true, cancel: "error" },
 		);
 	}
 
+	/**
+	 * Runs any steps before actual GC starts. It initializes state from base snapshot (required first time GC runs)
+	 * and notifies runtime that GC is starting.
+	 */
 	private async runPreGCSteps() {
```
[the remark this answers is not in the document]
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts]: done

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts]: Rephrased it.
I had initially deleted nodes from gcData directly (so had Tyler when he originally wrote this code). But it gives a tslint warning which made me think it's not safe to do so.

**Review by markfields — APPROVED** [review]
